package/android/cpp/rnskia-android/SkiaOpenGLHelper.h

```diff
@@ -0,0 +1,246 @@
+
+      if (context->directContext == nullptr) {
+        RNSkLogger::logToConsole("GrDirectContext::MakeGL failed");
+        return false;
+      }
+    }
+
+    // It all went well!
+    return true;
+  }
+
+private:
+  /**
+   * Initializes OpenGL using the provided SkiaOpenGLContext. This means
+   * getting the default display and calling eglInitializeContext. If the
+   * context already have a valid initialised display nothing will be done and
+   * the method will return true.
+   * @param context to store resulting eglDisplay in
+   * @return True if already initialized or if eglGetDisplay and eglInitialize
+   * was successfull.
+   */
+  static bool initializeOpenGL(SkiaOpenGLContext *context) {
+    if (context->glDisplay == EGL_NO_DISPLAY) {
+      // Get default display
+      context->glDisplay = eglGetDisplay(EGL_DEFAULT_DISPLAY);
+      if (context->glDisplay == EGL_NO_DISPLAY) {
+        RNSkLogger::logToConsole("eglGetDisplay failed : %i", glGetError());
+        return false;
+      }
+
+      EGLint major;
+      EGLint minor;
+
+      if (eglInitialize(context->glDisplay, &major, &minor) != EGL_TRUE) {
+        RNSkLogger::logToConsole("eglInitialize failed : %i", glGetError());
+        eglTerminate(context->glDisplay);
+        context->glDisplay = EGL_NO_DISPLAY;
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Creates a new GLContext. If the context has a valid context no new context
+   * will be created.
+   * @param context Context to save results in
+   * @param sharedContext Shared OpenGL Context
+   * @return True if the call to eglCreateContext returned a valid OpenGL
+   * Context or if the context already is setup.
+   */
+  static bool createOpenGLContext(SkiaOpenGLContext *context,
+                                  std::atomic<EGLContext> *sharedEglContext) {
+    // No need to create new EGL Context if it already exists.
+    if (context->glContext != EGL_NO_CONTEXT) {
+      return true;
+    }
+
+    // Create config
+    context->glConfig = getConfig(context->glDisplay);
+    if (context->glConfig == 0) {
+      return false;
+    }
+
+    // Create OpenGL context attributes
+    EGLint contextAttribs[] = {EGL_CONTEXT_CLIENT_VERSION, 2, EGL_NONE};
+
+    // Initialize the offscreen context for this thread
+    context->glContext = eglCreateContext(context->glDisplay, context->glConfig,
+                                          *sharedEglContext, contextAttribs);
+
+    if (context->glContext == EGL_NO_CONTEXT) {
+      RNSkLogger::logToConsole("eglCreateContext failed: %d\n", eglGetError());
+      return EGL_NO_CONTEXT;
+    }
+
+    // Save to shared context so that we can share data between threads in
+    // OpenGL. The first context created will be the one passed to any
+    // subsequent EGLContexts created.
+    if (*sharedEglContext == EGL_NO_CONTEXT) {
+      *sharedEglContext = context->glContext;
+    }
```

Comment:
This feels off because now you have a thread specfic context that is available on the other thread (as sharedElgContext).

If we were to draw a tree of Skia and open gl contextes and accross thread this would look strange. It would probably be easier to understand and less error prone to have a clean tree hierarchy (one context shared accross thread but not used directly in either thread and then each thread has one context that has the shared context as parent).

Reply:
It is declared as an std::atomic<EGLContext> which is thread safe - so it should be all good. You have mentioned the above a few times - I'm still for doing this lazy - meaning the first one loaded is the one used as the parent context.

Reply:
std::atomic ensures that only one thread can access the value at any given time - so it is completely thread safe. Uses some built in thread methods to be super fast.